pallets/asset/src/lib.rs

```diff
@@ -2494,4 +2516,60 @@ impl<T: Config> Module<T> {
             }
         })
     }
+
+    fn base_redeem_from_portfolio(
+        origin: T::Origin,
+        ticker: Ticker,
+        value: Balance,
+        agent_portfolio: PortfolioId,
```

Comment:
Most of this code can be shared with `base_redeem`.  If the `agent_portfolio` is changed to an `Option` type, then `None` can be used as the "Default portfolio".
```suggestion
        agent_portfolio: Option<PortfolioId>,
```

Reply:
```suggestion
        portfolio: PortfolioKind,
```
Using just `PortfolioKind` would make it much easier to share this `base_redeem` function with both `fn redeem` and `fn redeem_from_portfolio`.